### PR TITLE
Asynchronous Sink

### PR DESCRIPTION
**Pull request by 3Hren**
[ABI report](http://htmlpreview.github.io/?https://raw.githubusercontent.com/3Hren/blackhole/abi/b17eb9711dec646abd2fe83d4dd080c1c7e72e67-9c57ef424f097546c8a13558c659c299b07510fa.html): compatible.

### CODE REVIEW

**Comment by 3Hren** [comment]
@antmat r?

**Comment by 3Hren** [comment]
Still no review. Will be merged at Friday if there won't be any activity.

File: src/record.cpp [hunk 1]
```diff
@@ -38,6 +27,10 @@ record_t::record_t(severity_t severity,
     inner.attributes = attributes;
 }
 
+record_t::record_t(inner_t inner) noexcept {
```
Comment (antmat) [inline src/record.cpp]: Isn't it better to assign in initialization list?

File: include/blackhole/detail/sink/asynchronous.hpp [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+inline namespace v1 {
+namespace experimental {
+namespace sink {
+
+using detail::recordbuf_t;
+
+/// I can imagine: drop, sleep, wait.
+class overflow_policy_t {
+public:
+    enum class action_t {
```
Comment (antmat) [inline include/blackhole/detail/sink/asynchronous.hpp]: How does this used?
I see only interface and factory for policy accepting string

File: include/blackhole/detail/sink/asynchronous.hpp [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    /// This method is called when the queue is unable to enqueue more items. It's okay to throw
+    /// exceptions from here, they will be propagated directly to the sink caller.
+    virtual auto overflow() -> action_t = 0;
+
+    virtual auto wakeup() -> void = 0;
+};
+
+class overflow_policy_factory_t {
+public:
+    auto create(const std::string& name) const -> std::unique_ptr<overflow_policy_t>;
```
Comment (antmat) [inline include/blackhole/detail/sink/asynchronous.hpp]: A small notice of acceptable names would be great

File: include/blackhole/detail/sink/asynchronous.hpp [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+};
+
+class asynchronous_t : public sink_t {
+    struct value_type {
+        recordbuf_t record;
+        std::string message;
+    };
+
+    typedef cds::container::VyukovMPSCCycleQueue<value_type> queue_type;
```
Comment (antmat) [inline include/blackhole/detail/sink/asynchronous.hpp]: Is it ok to use all this stuff in public header?

File: include/blackhole/detail/sink/asynchronous.hpp [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+    std::unique_ptr<overflow_policy_t> overflow_policy;
+
+    std::thread thread;
+
+public:
+    asynchronous_t(std::unique_ptr<sink_t> wrapped, std::size_t factor = 10);
+
+    // TODO: Full customization.
+    asynchronous_t(std::unique_ptr<sink_t> sink,
```
Comment (antmat) [inline include/blackhole/detail/sink/asynchronous.hpp]: umm, this ctor includes all parameters from the previous one. Do you really need 2 different ctors?

File: include/blackhole/record.hpp [hunk 6]
```diff
@@ -15,12 +15,12 @@ using cpp17::string_view;
 
 class record_t {
 public:
+    struct inner_t;
```
Comment (antmat) [inline include/blackhole/record.hpp]: Why do require this to be public, as you only additionally use it in private ctor?

File: src/sink/asynchronous.p.cpp [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    typedef overflow_policy_t::action_t action_t;
+
+    mutable std::mutex mutex;
+    std::condition_variable cv;
+
+public:
+    virtual auto overflow() -> action_t {
+        std::unique_lock<std::mutex> lock(mutex);
+
+        // TODO: WTF? Replace with something more convenient and not filling my eyes with blood.
```
Comment (antmat) [inline src/sink/asynchronous.p.cpp]: Maybe 
http://en.cppreference.com/w/cpp/thread/yield or
http://en.cppreference.com/w/cpp/thread/sleep_for?

**Comment by 3Hren** [comment]
Cool! Thanks!

File: src/record.cpp (hunk 1 above)
Comment (3Hren) [inline src/record.cpp]: Here is an aligned storage cast and assignment.

File: include/blackhole/detail/sink/asynchronous.hpp (hunk 4 above)
Comment (3Hren) [inline include/blackhole/detail/sink/asynchronous.hpp]: It's detail.

File: include/blackhole/detail/sink/asynchronous.hpp (hunk 3 above)
Comment (3Hren) [inline include/blackhole/detail/sink/asynchronous.hpp]: It's detail, but: https://github.com/3Hren/blackhole/blob/c1e916d3027bf1aeb66751755936837a70701c23/include/blackhole/sink/asynchronous.hpp#L30

File: include/blackhole/detail/sink/asynchronous.hpp (hunk 5 above)
Comment (3Hren) [inline include/blackhole/detail/sink/asynchronous.hpp]: Well, this place is under construction (see commented parameters), but for MVP it's okay.

File: include/blackhole/record.hpp (hunk 6 above)
Comment (3Hren) [inline include/blackhole/record.hpp]: It's for owned record to be able to initialize this class directly from implementation type. Dark place, agree, but alternative solution involves friendship.

File: src/sink/asynchronous.p.cpp (hunk 7 above)
Comment (3Hren) [inline src/sink/asynchronous.p.cpp]: Here we have controlled sleep.

**Pull request merged**